src/java/org/apache/cassandra/utils/SortingIterator.java

```diff
@@ -0,0 +1,317 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+/**
+ * An iterator that lists a set of items in order.
+ * <p>
+ * This is intended for use where we would normally read only a small subset of the elements, or where we would skip
+ * over large sections of the sorted set. To implement this efficiently, we put the data in a binary heap and extract
+ * elements as the iterator is queried, effectively performing heapsort. We also implement a quicker skipTo operation
+ * where we remove all smaller elements and restore the heap for all of them in one step.
+ * <p>
+ * As in heapsort, the first stage of the process has complexity O(n), and every next item is extracted in O(log n)
+ * steps. skipTo works in O(m.log n) steps (where m is the number of skipped items), but is also limited to O(n) when m
+ * is large by the same argument as the initial heap construction.
+ * <p>
+ * The accepts and stores nulls as non-present values, which turns out to be quite a bit more efficient for iterating
```

Comment:
Something wrong with the grammar here at the beginning of this sentence. "The <what> accepts and stores nulls ..." ?

Reply:
Done